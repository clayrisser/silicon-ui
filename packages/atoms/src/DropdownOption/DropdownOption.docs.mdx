# Dropdown Option

> silicon ui dropdown option
