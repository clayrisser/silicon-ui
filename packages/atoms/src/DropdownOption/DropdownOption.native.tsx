import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/primitives';
import { TouchableOpacity, Picker as NativePicker } from 'react-native';
import { createStyled } from '../styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import {
  DropdownOptionProps,
  StyledDropdownOptionProps,
  splitProps
} from './dropdownOptionProps';

const StyledView: StyledComponent<
  StyledDropdownOptionProps,
  StyledDropdownOptionProps,
  any
> = styled.View(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const StyledNativePicker = createStyled<DropdownOptionProps>(NativePicker, [
  background,
  border,
  layout,
  position,
  shadow,
  space
]);

const NativeBasePicker: FC<DropdownOptionProps> = (
  props: DropdownOptionProps
) => {
  const color = useColor(props);
  const {
    customDropdownOptionProps,
    nativeDropdownOptionProps,
    styledDropdownOptionProps,
    touchableOpacityProps
  } = splitProps({
    ...props,
    color
  });
  const children =
    typeof customDropdownOptionProps.children === 'string' ? (
      <NativePicker.Item
        value={customDropdownOptionProps.children}
        label={customDropdownOptionProps.children}
      />
    ) : (
      customDropdownOptionProps.children
    );

  return (
    <StyledNativePicker
      {...styledDropdownOptionProps}
      {...nativeDropdownOptionProps}
    >
      {children}
    </StyledNativePicker>
  );
};

NativeBasePicker.defaultProps = {
  // fontFamily: 'body',
  // fontWeight: 'body',
  // activeOpacity: 1,
  // autoContrast: false,
  // backgroundColor: 'background',
  children: <></>
  // fontSize: 0,
  // lineHeight: 'body'
};

export default NativeBasePicker;
