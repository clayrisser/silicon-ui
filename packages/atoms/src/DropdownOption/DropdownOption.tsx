import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import {
  DropdownOptionProps,
  DetailedHTMLDropdownOptionProps,
  splitProps
} from './dropdownOptionProps';

const HTMLOption: StyledComponent<
  DetailedHTMLDropdownOptionProps,
  DropdownOptionProps,
  object
> = styled.option(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const DropdownOption: FC<DropdownOptionProps> = (
  props: DropdownOptionProps
) => {
  const color = useColor(props);
  const {
    styledDropdownOptionProps,
    customDropdownOptionProps,
    touchableOpacityProps
  } = splitProps({
    ...props,
    color
  });

  return (
    <HTMLOption
      {...customDropdownOptionProps}
      {...touchableOpacityProps}
      {...(styledDropdownOptionProps as any)}
    />
  );
};

DropdownOption.defaultProps = {
  activeOpacity: 1,
  backgroundColor: 'background',
  children: <></>,
  value: '',
  fontFamily: 'body',
  fontSize: 0,
  fontWeight: 'body',
  lineHeight: 'body'
};

export default styled(DropdownOption)`
  :active {
    opacity: ${({ activeOpacity }: DropdownOptionProps) => activeOpacity};
  }
`;
