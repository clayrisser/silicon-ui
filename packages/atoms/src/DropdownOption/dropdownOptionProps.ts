import { ReactNode, OptionHTMLAttributes, DetailedHTMLProps } from 'react';
import { TouchableOpacityProps, ViewProps } from 'react-native';
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLDropdownOptionProps = DetailedHTMLProps<
  OptionHTMLAttributes<HTMLOptionElement>,
  HTMLOptionElement
>;

export interface StyledDropdownOptionProps
  extends BackgroundProps,
    BorderProps,
    ColorProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {}

export interface NativeDropdownOptionProps extends ViewProps {
  value?: string;
  label?: string;
}

export interface CustomDropdownOptionProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  children?: ReactNode;
  theme?: Theme;
  value?: string;
  label?: string;
}

export interface DropdownOptionProps
  extends CustomDropdownOptionProps,
    NativeDropdownOptionProps,
    StyledDropdownOptionProps,
    TouchableOpacityProps {}

export const customDropdownOptionPropsKeys = new Set([
  'autoContrast',
  'children',
  'theme',
  'value',
  'label'
]);

export const touchableOpacityPropsKeys = new Set([
  'activeOpacity',
  'delayLongPress',
  'delayPressIn',
  'delayPressOut',
  'disabled',
  'hitSlop',
  'onBlur',
  'onFocus',
  'onLayout',
  'onLongPress',
  'pressRetentionOffset',
  'testID'
]);

export const nativeDropdownOptionPropsKeys = new Set<string>([
  'onValueChange',
  'value',
  'label'
  // 'onPress',
  // 'onPressIn',
  // 'onPressOut'
]);

export interface SplitProps {
  customDropdownOptionProps: CustomDropdownOptionProps;
  nativeDropdownOptionProps: NativeDropdownOptionProps;
  styledDropdownOptionProps: StyledDropdownOptionProps;
  touchableOpacityProps: TouchableOpacityProps;
}

export function splitProps(props: DropdownOptionProps): SplitProps {
  const styledDropdownOptionProps: { [key: string]: any } = {};
  const customDropdownOptionProps: { [key: string]: any } = {};
  const touchableOpacityProps: { [key: string]: any } = {};
  const nativeDropdownOptionProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customDropdownOptionPropsKeys.has(key)) {
      customDropdownOptionProps[key] = prop;
    } else if (touchableOpacityPropsKeys.has(key)) {
      touchableOpacityProps[key] = prop;
    } else if (nativeDropdownOptionPropsKeys.has(key)) {
      nativeDropdownOptionProps[key] = prop;
    } else {
      styledDropdownOptionProps[key] = prop;
    }
  });
  return {
    customDropdownOptionProps,
    nativeDropdownOptionProps,
    styledDropdownOptionProps,
    touchableOpacityProps
  };
}
