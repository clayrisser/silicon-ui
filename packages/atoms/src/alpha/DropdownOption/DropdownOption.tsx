import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import {
  DropdownOptionProps,
  DetailedHTMLDropdownOptionProps,
  splitProps
} from './dropdownOptionProps';

const HTMLOption: StyledComponent<
  DetailedHTMLDropdownOptionProps,
  DropdownOptionProps,
  object
> = styled.option(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const DropdownOption: FC<DropdownOptionProps> = (
  props: DropdownOptionProps
) => {
  const {
    styledDropdownOptionProps,
    customDropdownOptionProps,
    touchableOpacityProps
  } = splitProps(props);

  return (
    <HTMLOption
      {...customDropdownOptionProps}
      {...touchableOpacityProps}
      {...(styledDropdownOptionProps as any)}
    />
  );
};

DropdownOption.defaultProps = {
  backgroundColor: 'inverseText',
  children: <></>,
  value: '',
  fontFamily: 'body',
  fontSize: 0,
  fontWeight: 'body',
  lineHeight: 'body'
};

export default styled(DropdownOption)``;
