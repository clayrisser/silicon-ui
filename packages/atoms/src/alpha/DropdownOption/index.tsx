import DropdownOption from './DropdownOption';
import { DropdownOptionProps } from './dropdownOptionProps';

export { DropdownOptionProps };
export default DropdownOption;
