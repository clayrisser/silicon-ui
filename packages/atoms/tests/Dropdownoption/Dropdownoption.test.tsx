import React from 'react';
import renderer from 'react-test-renderer';
// import { render, fireEvent } from '@testing-library/react';
import DropdownOption from '../../src/DropdownOption';

describe('<DropdropOption />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<DropdownOption />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe('<DropdownOption withProps />', () => {
  it('renders  with all props correctly', () => {
    const value = 'Option';
    const display = 'Option';
    const disabled = true;
    const fontFamily = '';
    const backgroundColor = 'red';
    const fontWeight = 2;
    const fontSize = 3;
    const tree = renderer
      .create(
        <DropdownOption
          value={value}
          display={display}
          backgroundColor={backgroundColor}
          disabled={disabled}
          fontFamily={fontFamily}
          fontWeight={fontWeight}
          fontSize={fontSize}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
