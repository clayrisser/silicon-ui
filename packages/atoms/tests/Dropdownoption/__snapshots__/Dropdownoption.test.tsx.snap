// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownOption withProps /> renders  with all props correctly 1`] = `
<option
  className="css-1omk0he"
  disabled={true}
  display="Option"
  fontFamily=""
  fontSize={3}
  fontWeight={2}
  value="Option"
/>
`;

exports[`<DropdropOption /> renders correctly 1`] = `
<option
  className="css-viiq2s"
  fontFamily="body"
  fontSize={0}
  fontWeight="body"
  value=""
/>
`;
